cmd/registry/cmd/upload-csv.go

```diff
@@ -0,0 +1,256 @@
+	if err != nil {
+		return uploadCSVRow{}, err
+	}
+
+	// Build an index of { column name -> column number } using the header row.
+	if r.columnIndex == nil {
+		if err := r.buildColumnIndex(row); err != nil {
+			return uploadCSVRow{}, err
+		}
+
+		// Return the first non-header row instead of the header itself.
+		return r.Read()
+	}
+
+	return uploadCSVRow{
+		ApiID:     row[r.columnIndex["api_id"]],
+		VersionID: row[r.columnIndex["version_id"]],
+		SpecID:    row[r.columnIndex["spec_id"]],
+		Filepath:  row[r.columnIndex["filepath"]],
+	}, nil
+}
+
+func (r *uploadCSVReader) buildColumnIndex(header []string) error {
+	want := map[string]bool{
+		"api_id":     true,
+		"version_id": true,
+		"spec_id":    true,
+		"filepath":   true,
+	}
+
+	var errs []error
+	r.columnIndex = make(map[string]int, len(header))
+	for i, name := range header {
+		r.columnIndex[name] = i
+		if !want[name] {
+			errs = append(errs, fmt.Errorf("unexpected column name %q", name))
+		}
+	}
+
+	for name := range want {
+		if _, got := r.columnIndex[name]; !got {
+			errs = append(errs, fmt.Errorf("expected column name %q", name))
+		}
+	}
+
+	if len(errs) > 0 {
+		msg := fmt.Sprintf("invalid header %v:", header)
+		for _, err := range errs {
+			msg += "\n\t" + err.Error()
+		}
+		return errors.New(msg)
+	}
+
+	return nil
+}
+
+type uploadSpecTask struct {
+	ctx       context.Context
+	client    connection.Client
+	projectID string
+	apiID     string
+	versionID string
+	specID    string
+	filepath  string
+}
+
+func (t uploadSpecTask) Run() error {
+	api, err := t.client.CreateApi(t.ctx, &rpcpb.CreateApiRequest{
+		Parent: fmt.Sprintf("projects/%s", t.projectID),
+		ApiId:  t.apiID,
+		Api:    &rpcpb.Api{},
+	})
+
+	switch status.Code(err) {
+	case codes.OK:
+		log.Printf("Created API: %s", api.GetName())
+	case codes.AlreadyExists:
+		api = &rpc.Api{
+			Name: fmt.Sprintf("projects/%s/apis/%s", t.projectID, t.apiID),
+		}
+	default:
+		return fmt.Errorf("failed to ensure API exists: %s", err)
+	}
+
+	version, err := t.client.CreateApiVersion(t.ctx, &rpcpb.CreateApiVersionRequest{
+		Parent:       api.GetName(),
+		ApiVersionId: t.versionID,
+		ApiVersion:   &rpcpb.ApiVersion{},
+	})
+
+	switch status.Code(err) {
+	case codes.OK:
+		log.Printf("Created API version: %s", version.GetName())
+	case codes.AlreadyExists:
+		version = &rpc.ApiVersion{
+			Name: fmt.Sprintf("projects/%s/apis/%s/versions/%s", t.projectID, t.apiID, t.versionID),
+		}
+	default:
+		return fmt.Errorf("failed to ensure API version exists: %s", err)
+	}
+
+	contents, err := ioutil.ReadFile(t.filepath)
+	if err != nil {
+		return err
+	}
+
+	compressed, err := core.GZippedBytes(contents)
+	if err != nil {
+		return err
+	}
+
+	spec, err := t.client.CreateApiSpec(t.ctx, &rpcpb.CreateApiSpecRequest{
+		Parent:    version.GetName(),
+		ApiSpecId: t.specID,
+		ApiSpec: &rpcpb.ApiSpec{
+			// TODO: How do we choose a mime type?
+			MimeType: core.OpenAPIMimeType("+gzip", "3.0.0"),
```

Comment:
I'm not sure how to approach this. Should we make a guess by checking if the file is "openapi.yaml" or ends with ".proto" and leave it blank if we're not sure?